pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,385 @@
+// The function is stateless and idempotent. If the error was returned before all nodes' states were processed,
+// ApplyState would be called again and complete the processing - all the decisions are based on the input data.
+func (m *ClusterUpgradeStateManager) ApplyState(ctx context.Context,
+	currentState *ClusterUpgradeState, upgradePolicy *v1alpha1.OfedUpgradePolicySpec) error {
+	m.Log.V(consts.LogLevelInfo).Info("State Manager, got state update")
+
+	if currentState == nil {
+		return fmt.Errorf("currentState should not be empty")
+	}
+
+	if upgradePolicy == nil || !upgradePolicy.AutoUpgrade {
+		m.Log.V(consts.LogLevelInfo).Info("Driver auto upgrade is disabled, skipping")
+		return nil
+	}
+
+	m.Log.V(consts.LogLevelInfo).Info("Node states:",
+		"Unknown", len(currentState.NodeStates[UpgradeStateUnknown]),
+		UpgradeStateDone, len(currentState.NodeStates[UpgradeStateDone]),
+		UpgradeStateUpgradeRequired, len(currentState.NodeStates[UpgradeStateUpgradeRequired]),
+		UpgradeStateDrain, len(currentState.NodeStates[UpgradeStateDrain]),
+		UpgradeStateDrainFailed, len(currentState.NodeStates[UpgradeStateDrainFailed]),
+		UpgradeStatePodRestart, len(currentState.NodeStates[UpgradeStatePodRestart]))
+
+	upgradesInProgress := len(currentState.NodeStates[UpgradeStateDrain]) +
+		len(currentState.NodeStates[UpgradeStatePodRestart]) +
+		len(currentState.NodeStates[UpgradeStateDrainFailed]) +
+		len(currentState.NodeStates[UpgradeStateUncordonRequired])
+
+	var upgradesAvailable int
+	if upgradePolicy.MaxParallelUpgrades == 0 {
+		// Only nodes in UpgradeStateUpgradeRequired can start upgrading, so all of them will move to drain stage
+		upgradesAvailable = len(currentState.NodeStates[UpgradeStateUpgradeRequired])
+	} else {
+		upgradesAvailable = upgradePolicy.MaxParallelUpgrades - upgradesInProgress
+	}
+
+	m.Log.V(consts.LogLevelInfo).Info("Upgrades in progress",
+		"currently in progress", upgradesInProgress,
+		"max parallel upgrades", upgradePolicy.MaxParallelUpgrades,
+		"upgrade slots available", upgradesAvailable)
+
+	// First, check if unknown or ready nodes need to be upgraded
+	err := m.ProcessDoneOrUnknownNodes(ctx, currentState, UpgradeStateUnknown)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to process nodes", "state", UpgradeStateUnknown)
+		return err
+	}
+	err = m.ProcessDoneOrUnknownNodes(ctx, currentState, UpgradeStateDone)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to process nodes", "state", UpgradeStateDone)
+		return err
+	}
+	// Start upgrade process for upgradesAvailable number of nodes
+	err = m.ProcessUpgradeRequiredNodes(ctx, currentState, upgradesAvailable)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(
+			err, "Failed to process nodes", "state", UpgradeStateUpgradeRequired)
+		return err
+	}
+	// Schedule nodes for drain
+	err = m.ProcessDrainNodes(ctx, currentState, upgradePolicy.DrainSpec)
```

Comment:
The reconcile loop would probably be triggered by node/ds update but for safety I added a periodic reconcile every 2 minutes with an explaining comment in `controllers/upgrade_controller.go`

Reply:
ds will not update in this case, we will be depentent on external triggers.

thx for adding the periodic reconcile. i think it may also solve issues that were not found in the review :)